### DISCUSSION TITLE
How to use lifetime in event?

### DISCUSSION THREAD

**Discussion by PhanDungTri** [bevyengine/bevy#5211] [Q&A]
Hi, I'm a beginner in Rust and Bevy. I can basically understand the concept of 'lifetime' but still can't actually use it in practice.

Here is my situation, I have defined a struct that has a property referencing to another struct.

```rust
pub struct JobProcessingEvent<'a> {
  pub job: &'a Job,
  pub effort: f32,
}
```

I want to use it in a system but don't know how to specify those lifetimes.

```rust
pub fn process_job_system(
  job_completed_event: EventWriter<JobProcessingEvent>,
  mut commands: Commands,
  peasant_query: Query<(Entity, &Job)>,
) {
  //...
}
```

Any guidance is very appreciated. Thank you.

**Comment by james7132** [comment]
[Events must be `'static`.](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/event.rs#L17=). They cannot have non-`'static` fields. You'll likely need to find other ways to refer to the Job (i.e. a JobId or Entity) instead of a reference.

**Comment by nicopap** [comment]
You can't pass a reference to a component through an event. This is because the lifetime of `&Job` in `process_job_system` is limited to that system. This is because at any time when `process_job_system` is not running, another system can update the value pointed-to by the reference, and this violates rust's aliasing rules. If you add a reference to `Job` in an event, it would require that the lifetime of `&Job` goes beyond `process_job_system`, which we already said is not possible.

This needs a different solution. How does your `Job` struct looks like? Most likely, you want your `JobProcessingEvent` to own a value of `Job`, or maybe an ID for it. If you decide to own the `Job`, you need to add `#[derive(Clone)]` to your `Job` declaration and use `job.clone()` to copy it into the event.
